File: src/main/java/me/delivery/domain/address/user/model/entity/UserAddress.java
package me.delivery.domain.address.user.model.entity;

import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import me.delivery.config.exception.ForbiddenException;
import me.delivery.domain.address.model.entity.Address;
import me.delivery.domain.address.model.vo.AddressVO;
import me.delivery.domain.address.user.model.error_code.UserAddressErrorCode;
import me.delivery.domain.address.user.model.vo.UserAddressVO;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;

import javax.persistence.*;
import java.time.LocalDateTime;

@Entity
@Getter
@Setter
@NoArgsConstructor
public class UserAddress {
    // id, createdAt - baseEntity를 상속후 삭제될 예정
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;
    @CreatedDate
    private LocalDateTime createdAt;

    private long addressId;
    private long userId;
    private String detailAddress;
    private String memo;
    @LastModifiedDate
    private LocalDateTime updatedAt;
    private LocalDateTime deletedAt;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "addressId", insertable = false, updatable = false)
    private Address address;

    @Builder
    private UserAddress(long addressId, long userId, String detailAddress, String memo) {
        this.addressId = addressId;
        this.detailAddress = detailAddress;
        this.memo = memo;
    }

    public UserAddressVO toImmutable(AddressVO addressVO) {
        return UserAddressVO.builder()
                .id(id)
                .createdAt(createdAt)
                .memo(memo)
                .address(addressVO)
                .build();
    }

    public void delete(long userId) {
        if (this.userId != userId) {
            throw new ForbiddenException(UserAddressErrorCode.DeleteForbidden);
        }

        deletedAt = LocalDateTime.now();
    }

    public void updateMemoAndDetailAddress(int userId, String memo, String detailAddress) {
        if (this.userId != userId) {
            throw new ForbiddenException(UserAddressErrorCode.UpdateForbidden);
        }

        this.memo = memo;
        this.detailAddress = detailAddress;
    }
}

File: src/main/kotlin/me/delivery/domain/address/user/model/entity/UserAddress.kt
package me.delivery.domain.address.user.model.entity

import me.delivery.config.exception.ForbiddenException
import me.delivery.domain.address.model.entity.Address
import me.delivery.domain.address.model.vo.AddressVO
import me.delivery.domain.address.user.model.error_code.UserAddressErrorCode
import me.delivery.domain.address.user.model.vo.UserAddressVO
import me.delivery.domain.entity.BaseEntity
import org.springframework.data.annotation.LastModifiedDate
import java.time.LocalDateTime
import javax.persistence.Entity
import javax.persistence.FetchType
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne

@Entity
class UserAddress(
    val addressId: Long,
    val userId: Long,
    var detailAddress: String,
    var memo: String? = null,
    @LastModifiedDate
    var updatedAt: LocalDateTime? = null,
    var deletedAt: LocalDateTime? = null,
): BaseEntity() {

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "addressId", insertable = false, updatable = false)
    lateinit var address: Address

    fun toImmutable(addressVO: AddressVO) = UserAddressVO(this)

    fun delete(userId: Long) {
        if (this.userId != userId) {
            throw ForbiddenException(UserAddressErrorCode.DeleteForbidden)
        }
        deletedAt = LocalDateTime.now()
    }

    fun updateMemoAndDetailAddress(userId: Int, memo: String, detailAddress: String) {
        if (this.userId != userId.toLong()) {
            throw ForbiddenException(UserAddressErrorCode.UpdateForbidden)
        }
        this.memo = memo
        this.detailAddress = detailAddress
    }
}